cli/ts/commands/joinPoll.ts

```diff
@@ -0,0 +1,386 @@
+  const inputHash = sha256Hash([nullifier, credits, stateRoot, pollPubKeyArray[0], pollPubKeyArray[1]]);
+
+  const circuitInputs = {
+    privKey: maciPrivKey.asCircuitInputs(),
+    pollPrivKey: pollPrivKey.asCircuitInputs(),
+    pollPubKey: pollPubKey.asCircuitInputs(),
+    stateLeaf: stateLeafArray,
+    siblings: siblingsArray,
+    indices,
+    nullifier,
+    credits,
+    stateRoot,
+    actualStateTreeDepth,
+    inputHash,
+  };
+
+  return stringifyBigInts(circuitInputs) as unknown as IPollJoiningCircuitInputs;
+};
+
+export const joinPoll = async ({
+  maciAddress,
+  privateKey,
+  pollPrivKey,
+  stateIndex,
+  newVoiceCreditBalance,
+  stateFile,
+  pollId,
+  signer,
+  startBlock,
+  endBlock,
+  blocksPerBatch,
+  transactionHash,
+  pollJoiningZkey,
+  useWasm,
+  rapidsnark,
+  pollWitgen,
+  pollWasm,
+  quiet = true,
+}: IJoinPollArgs): Promise<number> => {
+  banner(quiet);
+
+  if (!(await contractExists(signer.provider!, maciAddress))) {
+    logError("MACI contract does not exist");
+  }
+
+  if (!PrivKey.isValidSerializedPrivKey(privateKey)) {
+    logError("Invalid MACI private key");
+  }
+
+  const userMaciPrivKey = PrivKey.deserialize(privateKey);
+  const userMaciPubKey = new Keypair(userMaciPrivKey).pubKey;
+  const nullifier = poseidon([BigInt(userMaciPrivKey.asCircuitInputs())]);
+
+  // Create poll public key from poll private key
+  const pollPrivKeyDeserialized = PrivKey.deserialize(pollPrivKey);
+  const pollKeyPair = new Keypair(pollPrivKeyDeserialized);
+  const pollPubKey = pollKeyPair.pubKey;
+
+  if (pollId < 0) {
+    logError("Invalid poll id");
+  }
+
+  const maciContract = MACIFactory.connect(maciAddress, signer);
+  const pollAddress = await maciContract.getPoll(pollId);
+
+  if (!(await contractExists(signer.provider!, pollAddress))) {
+    logError("Poll contract does not exist");
+  }
+
+  const pollContract = PollFactory.connect(pollAddress, signer);
+
+  let loadedStateIndex = stateIndex;
+  let loadedCreditBalance = newVoiceCreditBalance;
+
+  let maciState: MaciState | undefined;
+  let signUpData: IGenSignUpTree | undefined;
+  let currentStateRootIndex: number;
+  let circuitInputs: CircuitInputs;
+  if (stateFile) {
+    try {
+      const file = await fs.promises.readFile(stateFile);
+      const content = JSON.parse(file.toString()) as unknown as IJsonMaciState;
+      maciState = MaciState.fromJSON(content);
+    } catch (error) {
+      logError((error as Error).message);
+    }
+    const poll = maciState!.polls.get(pollId)!;
+
+    if (poll.hasJoined(nullifier)) {
+      throw new Error("User the given nullifier has already joined");
+    }
+
+    if (stateIndex) {
+      const index = maciState?.stateLeaves.findIndex((leaf) => leaf.pubKey.equals(userMaciPubKey));
+      if (index) {
+        loadedStateIndex = BigInt(index);
+      } else {
+        logError("State leaf not found");
+        process.exit();
+      }
+    }
```

Comment:
There are too many nested if/else + try/catch. It's better to refactor this code to improve readability.
It's related to all the cases for this file.